### PR TITLE
Backport for using lastSyncedGlobalCheckpoint in deletion policy

### PR DESCRIPTION
**Pull request by dnhatn**
This is a backport of #27826. 

@bleskes, Could you please take a quick look to make sure that this backport fit with your fix in https://github.com/elastic/elasticsearch/pull/26734.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 1]
```diff
@@ -446,33 +447,27 @@ private void recoverFromTranslogInternal() throws IOException {
...
-                    IOUtils.closeWhileHandlingException(translog);
-                }
-            }
+        return new Translog(translogConfig, translogUUID, translogDeletionPolicy, globalCheckpointSupplier);
+    }
+
+    private void refreshTranslogUUIDIfNeeded(IndexWriter writer, Translog translog) throws IOException {
+        final String translogUUID = commitDataAsMap(writer).get(Translog.TRANSLOG_UUID_KEY);
+        if (translog.getTranslogUUID().equals(translogUUID) == false) {
+            assert openMode != EngineConfig.OpenMode.OPEN_INDEX_AND_TRANSLOG : "OPEN_INDEX_AND_TRANSLOG must not fresh translogUUID";
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: assertion message reads funny

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 2]
```diff
@@ -195,10 +198,9 @@ public InternalEngine(EngineConfig engineConfig) {
                 historyUUID = loadOrGenerateHistoryUUID(writer, engineConfig.getForceNewHistoryUUID());
                 Objects.requireNonNull(historyUUID, "history uuid should not be null");
                 indexWriter = writer;
-                translog = openTranslog(engineConfig, writer, translogDeletionPolicy, () -> seqNoService.getGlobalCheckpoint());
-                assert translog.getGeneration() != null;
                 // we can only do this after we generated and committed a translog uuid. other wise the combined
                 // retention policy, which listens to commits, gets all confused.
+                refreshTranslogUUIDIfNeeded(writer, translog);
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: move this line above the comment? The comment applies to persistHistoryUUIDIfNeeded

**Review by ywelsch — COMMENTED** [review]
I want @bleskes to have a look too.

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 2 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: I prefer it like this, wdyt?

```
diff --git a/core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java b/core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java
index b0523a1c3d2..59ddbb82ccb 100644
--- a/core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java
+++ b/core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java
@@ -198,10 +198,11 @@ public class InternalEngine extends Engine {
                 historyUUID = loadOrGenerateHistoryUUID(writer, engineConfig.getForceNewHistoryUUID());
                 Objects.requireNonNull(historyUUID, "history uuid should not be null");
                 indexWriter = writer;
-                // we can only do this after we generated and committed a translog uuid. other wise the combined
-                // retention policy, which listens to commits, gets all confused.
-                refreshTranslogUUIDIfNeeded(writer, translog);
-                persistHistoryUUIDIfNeeded();
+                if (openMode == EngineConfig.OpenMode.OPEN_INDEX_AND_TRANSLOG) {
+                    persistHistoryUUIDIfNeeded();
+                } else {
+                    commitNewTranslogAndHistoryUUIDs(writer, translog);
+                }
             } catch (IOException | TranslogCorruptedException e) {
                 throw new EngineCreationFailureException(shardId, "failed to create engine", e);
             } catch (AssertionError e) {
@@ -461,13 +462,15 @@ public class InternalEngine extends Engine {
         return new Translog(translogConfig, translogUUID, translogDeletionPolicy, globalCheckpointSupplier);
     }

-    private void refreshTranslogUUIDIfNeeded(IndexWriter writer, Translog translog) throws IOException {
-        final String translogUUID = commitDataAsMap(writer).get(Translog.TRANSLOG_UUID_KEY);
-        if (translog.getTranslogUUID().equals(translogUUID) == false) {
-            assert openMode != EngineConfig.OpenMode.OPEN_INDEX_AND_TRANSLOG : "OPEN_INDEX_AND_TRANSLOG must not fresh translogUUID";
-            commitIndexWriter(writer, translog, openMode == EngineConfig.OpenMode.OPEN_INDEX_CREATE_TRANSLOG
-                ? commitDataAsMap(writer).get(SYNC_COMMIT_ID) : null);
-        }
+    private void commitNewTranslogAndHistoryUUIDs(IndexWriter writer, Translog translog) throws IOException {
+        final Map<String, String> commitUserData = commitDataAsMap(writer);
+        assert commitUserData.get(Translog.TRANSLOG_UUID_KEY).equals(translog.getTranslogUUID()) == false :
+            "committing a new translog uuid but it's already equal (uuid [" + translog.getTranslogUUID() + "](";
+        assert openMode != EngineConfig.OpenMode.OPEN_INDEX_AND_TRANSLOG : "OPEN_INDEX_AND_TRANSLOG must not fresh translogUUID";
+        assert historyUUID.equals(commitUserData.containsKey(HISTORY_UUID_KEY)) == false || config().getForceNewHistoryUUID() == false :
+            "a new history id was forced, but the one in the commit is equal";
+        commitIndexWriter(writer, translog, openMode == EngineConfig.OpenMode.OPEN_INDEX_CREATE_TRANSLOG
+            ? commitUserData.get(SYNC_COMMIT_ID) : null);
     }
```

**Comment by dnhatn** [comment]
please retest this.

**Comment by dnhatn** [comment]
@ywelsch, This backport is tricky than we thought. Sorry for blocking your PR.

**Review by bleskes — APPROVED** [review]
LGTM

**Comment by bleskes** [comment]
once the ci is green, this can go in.

**Comment by dnhatn** [comment]
Thanks @bleskes for reviewing.

**Pull request merged**